src/Npgsql/NpgsqlConnector.Auth.cs

```diff
@@ -90,29 +93,89 @@ async Task AuthenticateSASL(List<string> mechanisms, bool async)
+            var saltBytes = Convert.FromBase64String(firstServerMsg.Salt);
+            var saltedPassword = Hi(passwd.Normalize(NormalizationForm.FormKC), saltBytes, firstServerMsg.Iteration);
+
+            var clientKey = HMAC(saltedPassword, ClientKey);
+            byte[] storedKey;
+            using (var sha256 = SHA256.Create())
+                storedKey = sha256.ComputeHash(clientKey);
+
+            var clientFirstMessageBare = "n=*,r=" + clientNonce;
+            var serverFirstMessage = $"r={firstServerMsg.Nonce},s={firstServerMsg.Salt},i={firstServerMsg.Iteration}";
+            var clientFinalMessageWithoutProof = "c=biws,r=" + firstServerMsg.Nonce;
+
+            var authMessage = $"{clientFirstMessageBare},{serverFirstMessage},{clientFinalMessageWithoutProof}";
+
+            var clientSignature = HMAC(storedKey, authMessage);
+            var clientProofBytes = XOR(clientKey, clientSignature);
+            var clientProof = Convert.ToBase64String(clientProofBytes);
+
+            var serverKey = HMAC(saltedPassword, ServerKey);
+            var serverSignature = HMAC(serverKey, authMessage);
+
+            var messageStr = $"{clientFinalMessageWithoutProof},p={clientProof}";
+
+            await WriteSASLResponse(Encoding.UTF8.GetBytes(messageStr), async);
+            await Flush(async);
 
             var saslFinalServerMsg = Expect<AuthenticationSASLFinalMessage>(await ReadMessage(async));
             if (saslFinalServerMsg.AuthRequestType != AuthenticationRequestType.AuthenticationSASLFinal)
                 throw new NpgsqlException("[SASL] AuthenticationSASLFinal message expected");
             var scramFinalServerMsg = new AuthenticationSCRAMServerFinalMessage(saslFinalServerMsg.Payload);
 
-            if (scramFinalServerMsg.ServerSignature != Convert.ToBase64String(scramFinalClientMsg.ServerSignature))
+            if (scramFinalServerMsg.ServerSignature != Convert.ToBase64String(serverSignature))
                 throw new NpgsqlException("[SCRAM] Unable to verify server signature");
 
             var okMsg = Expect<AuthenticationRequestMessage>(await ReadMessage(async));
             if (okMsg.AuthRequestType != AuthenticationRequestType.AuthenticationOk)
                 throw new NpgsqlException("[SASL] Expected AuthenticationOK message");
 
-            string GetNonce()
+            static string GetNonce()
             {
                 var nonceLength = 18;
-                var rncProvider = RandomNumberGenerator.Create();
-                var nonceBytes = new byte[nonceLength];
-                rncProvider.GetBytes(nonceBytes);
-                return Convert.ToBase64String(nonceBytes);
+                using (var rncProvider = RandomNumberGenerator.Create())
+                {
+                    var nonceBytes = new byte[nonceLength];
+                    rncProvider.GetBytes(nonceBytes);
+                    return Convert.ToBase64String(nonceBytes);
+                }
```

Comment:
Oh, dear, why not spans?

```suggestion
                Span<byte> nonceBytes = stackalloc byte[18];
                RandomNumberGenerator.Fill(nonceBytes);
                return Convert.ToBase64String(nonceBytes);
```

Reply:
Just copy-pasting :) It's really not perf-sensitive code (only used when physically connecting), and the `Fill()` method exists only in netstandard2.1. We could condition on the TFM, but frankly it doesn't seem worth it (for this specific case!). Do you agree?

Reply:
Okay, but could you replace the `nonceLength` variable by `const` or just remove it?

Reply:
Removed it, makes sense.